### ISSUE TITLE
Benchmark in readme is really too misleading

### ISSUE THREAD

**Issue by gpnuma** [Cyan4973/lz4#109]
*Disclaimer : I'm the author of [density](https://github.com/centaurean/density.git)*

I know benchmarks are always a subject for debate, that's why I'm not really bothered when results vary from platform to platform, compiler to compiler or even between file types, but in this case I honestly think that your benchmark is grossly misleading, and for at least 2 reasons :

**It is incomplete**
What I don't understand is that you use fsbench (without providing the source code of the lz4 library version you're using) which includes much more codecs. Where are LZF, wfLZ, and [density](https://github.com/centaurean/density.git) for example ? It's nice to add zlib as a reference but is it really useful ? That's not really what lz4 competes against.
If you want an up-to-date version of fsbench with source code of the codecs I maintain one [here](https://github.com/gpnuma/fsbench-density.git).

Also, you only benchmark the file *silesia*, and you omit text files like *enwik8* or database files etc. In my case (density) people were concerned that I benchmarked against *enwik8* only so I added *silesia*.

I know you're aware of the fantastic benchmark by @nemequ which can be seen [here](https://quixdb.github.io/squash-benchmark/?dataset=cp.html&machine=s-desktop), it's probably the best there is today (lots of platforms, files, codecs)... why is there no link to it on your project's page ?

All of this gives the impression that lz4 is in a class of its own when one considers high-speed compression, but that is clearly not the case if you actually compare it to libraries designed with the same purpose in mind.

**It is biased**
That's actually a direct consequence of the previous problem. I suspect for example that lz4 was optimized against *silesia*, there's nothing wrong with that as long as you also benchmark it against something for which it wasn't optimized (*enwik8* ?). Simple use case : when you run "lz4 fast 17" with *enwik8* the compression ratio is terrible.

Another example : I think (maybe I'm mistaken) that you use lz4_decompress_fast in your tests, which is inherently unsafe whereas other libraries' decompress functions are safe... it's not really fair to compare apples with oranges.

Well that's it, enough ranting for me :wink: , but it really is disappointing to see this kind of things on such a classy project like lz4.

**Comment by nemequ** [comment]
It would be inappropriate for me to get too involved with this issue, but I would like to make a couple quick points:

IMHO Silesa is a better dataset than enwik8.  I don't think people often need to compress 100 megabytes of plain text (other than log files, which enwik8 still isn't a good representation of).  Even given fsbench's somewhat odd methodology (IIRC it truncates each file to a specific size, I think 1 megabyte, then concatenates them all together and compresses the result) I think it's more informative than enwik8.

I think including zlib is a very good idea.  It's pretty much the standard for data compression, so it makes for a good reference point, especially for people less familiar with compression algorithms.

I can't figure out how to phrase anything else in a way that doesn't sound very self-serving, so I'm going to stop there.

**Comment by Cyan4973** [comment]
Even if the initial tone is more of a rant, there are a couple of interesting points that deserve some constructive answer, if only to provide useful information to readers interested in data compression.
The only issue is that it requires time to answer properly, so I'll get back to this thread as soon as I get some.

**Comment by gpnuma** [comment]
@nemequ I completely agree that *silesia* is more "complete" than *enwik8* in terms of data variety, but I find it interesting that more and more network protocols rely on the use of JSON, which happens to resemble *enwik8* very closely.
Also, most of the data created by computer users is either audio, video, photos or text, the first 3 are clearly out of the scope for general use-case compressors, so the only remaining one we can work on is text - and I think it's the most complex one, because text is a creation of man and fantastic text compression would require an extremely clever AI, to properly abstract multiple languages' grammar and semantics.
All of this to say that text shouldn't be discarded in the middle of gigabytes of binary data, as binary data formats will change but ASCII or Unicode text encoding will probably stay for a very, very long time.
As a side note, I favor your [new corpus](https://github.com/nemequ/squash-corpus) project, if everyone can agree on what an "average" dataset looks like that would be a good start, although, very commonly developers will choose a library which is best tailored for their needs instead of the best allround performer.

@Cyan4973 Sorry if the tone is not very appropriate I apologize for that, there's nothing personal I honestly appreciate your work and the work of all the other devs of LZ4, I just tend to get annoyed because a lot of library developers (whatever the field, not necessarily compression) just constantly manipulate benchmarks for their own needs and therefore trick users.
I'm not saying it's a "conscious" habit, it probably isn't, but I was hoping that a project with the stature of LZ4 would avoid such pitfalls.
Of course if you want I can re-write my points in a new issue without any "ranting".

**Comment by nemequ** [comment]
> but I find it interesting that more and more network protocols rely on the use of JSON, which happens to resemble enwik8 very closely.

I'm not sure enwik8 is a very good test for JSON, though.  Sure they are both text, but enwik8 is pretty different.  Obviously there is the size difference which, beyond the issue of longer distance matches being possible, also skews the results away from algorithms with quick start-up times.  JSON is probably also more compressible (e.g., something with an array of objects with identical keys and likely similar values).

> As a side note, I favor your new corpus project, if everyone can agree on what an "average" dataset looks like that would be a good start, although, very commonly developers will choose a library that is best tailored for their needs instead of the best allround performer.

I'm not planning on designing the new corpus with the idea of being able to generate a single number which makes sense for all-around performance since I don't think it's possible (see [the "overall corpus design" blurb in the squash-corpus README](https://github.com/nemequ/squash-corpus#non-goal-overall-corpus-design)).  Everyone's data is different, and I think people *should* choose something which is tailored for their use case…  I guess the closest thing to "general purpose" I'm planning to include would be nemequ/squash-corpus#4 (a disk image of a Linux distro), though I'm happy to listen to any other ideas.

Anyways, I'm getting a bit off-topic here.  I'm happy to discuss this, but the squash-corpus issue tracker would probably be more a more appropriate place.

IMHO any benchmark with a small enough amount of data to fit well in the README is going to be misleading for a lot of people.  I think the only way to mitigate that is to provide a lot more data points so people can select the data which is relevant for their use case, but unfortunately that can be more difficult for people to grok.

**Comment by gpnuma** [comment]
> I'm not sure enwik8 is a very good test for JSON, though. Sure they are both text, but enwik8 is pretty different. Obviously there is the size difference which, beyond the issue of longer distance matches being possible, also skews the results away from algorithms with quick start-up times. JSON is probably also more compressible (e.g., something with an array of objects with identical keys and likely similar values).

It probably is, the two have common properties though, in the fact that they are text based and their symbol distribution frequency table should be similar, although I'm giving it a guess here.

> IMHO any benchmark with a small enough amount of data to fit well in the README is going to be misleading for a lot of people

Now on this I quite disagree, it is very simple to add a line with another codec, especially when the codec in question is a direct alternative to your library.

Just take a look at https://github.com/Cyan4973/lz4/commit/979a991badffa8359232a6b7b74756a9821d2053 , you'll see that it has actually been the case, but this was rolled back for some unknown reason.

It's also possible to argue that *LZ4 fast* was developed as a response to density's higher speed algorithms, as they were in a class of their own before the LZ4 r129 release (RLE64's compression ratio isn't good enough on general purpose files), which makes this even more odd.

To summarize, personally I don't see any good reason for not showing other alternatives (and I'm not talking about density only), as that's what I would expect from a good open source project : openness and information.

**Comment by gpnuma** [comment]
I see that a LZF line was added in 7d72c0cb546435d5b1ab96c7513bdb0f2f6fceba, that's great news !

**Comment by nemequ** [comment]
> It probably is, the two have common properties though, in the fact that they are text based and their symbol distribution frequency table should be similar, although I'm giving it a guess here.

Me too.  I really need to get that new corpus finished so we can stop guessing… ;)

> Now on this I quite disagree, it is very simple to add a line with another codec, especially when the codec in question is a direct alternative to your library.

The problem is where to stop.  Squash has 40 codecs in 29 plugins and that's just a fraction of what is out there, especially if you count the proprietary stuff.  A lot of them (like density and lz4) have multiple compression levels, and many even have other ways to tune the algorithms (e.g., zlib has window bits, memory level, and strategy), and of course performance can vary a lot by architecture.  Some of them can perform much better with different compilers.  There are also considerations other than speed, such as memory usage.

It's simply not feasible to put all that data in the README—you need to draw a line somewhere, but you have to recognize that wherever you draw that line it's going to be misleading for some people… that's just the nature of incomplete data.

**Comment by gpnuma** [comment]
> The problem is where to stop.

Yes and no... there aren't many open source libraries that "compete" in the same areas.
For me things are very simple in the case of LZ4 :

LZ4 | "Competitor"
--- | ---
LZ4 fast |  Density chameleon
LZ4 | Density cheetah, LZF, LZO, QuickLZ, Snappy, wfLZ
In-between LZ4 and LZ4 HC | Density lion
LZ4 HC | ZSTD, zlib

Overall I've counted 10 "competitor" algorithms ! There might be some others I can't recall but 10-15 lines can fit easily in a project description IMHO.
If you add a couple of lines with links to other third-party benchmarks like yours @nemequ , there's still plenty of space left !

I'm pretty sure @Cyan4973 knows all of this, I'm just wondering why it doesn't appear in the LZ4 readme... maybe there's a good reason for it, it would be nice if we could have an answer from him or at least from someone in the LZ4 team.

In any case, if anyone is interested by a comparison between LZ4, density, and many other algorithms I invite you to take a look at [density's homepage](https://github.com/centaurean/density.git) for a much broader set of links and tables than you find here.

**Comment by Cyan4973** [comment]
OK, real life is really unforgiving these days, leaving only very little availability. Nonetheless, time to get into some answer.

The content is large, so I'll divide it into several paragraphs, as follows : 

- Choosing a Reference Data corpus
- (TBD later)
- (TBD later)

**Comment by Cyan4973** [comment]
Selecting a Reference data corpus
---------------------------------------------------

> It is biased. You omit text files like enwik8 or database files etc. In my case (density) people were concerned that I benchmarked against enwik8 only so I added silesia. I suspect for example that lz4 was optimized against silesia,

The question of the **reference compression corpus** is a frequent one, and deserves some explanation.

There is no "perfect" data compression corpus. Whatever is being tested, there is always someone which can complain that it's own use case is different, or insufficiently representative. And that's likely true, given the variety of situations.

The previous decade has seen the rise of `enwik8` as a reference corpus. It is promoted by Matt Mahoney, basically the most senior and famous expert in lossless data compression, author of world-record PAQ serie. It has a [prize](http://prize.hutter1.net/). So of course, it is tempting to use this file as a respectable reference.

But unfortunately, it would prove a bad choice.
And the reason is pretty simple : `enwik8` is _unlike_ most compression workload.

`enwik` is a specific corner case, mostly representative of its category. Human-redacted text files (Markdown format and such) belong to it too. But their proportion relative to the mass of data to compress is relatively low these days.

Don't confuse above definition with "human readable" files. Log files, xml schematics, html, soap, even source code are readable, but generate some very different redundancy patterns. Their semantic is much more "predictable", resulting in much better and longer hit ratio. Indeed, such files are much more common nowadays, so they matter much more than "raw human-redacted text files".

To get a sense of this difference, I recommend using the program `fv` from Matt Mahoney (accessible [here](http://mattmahoney.net/dc/textdata.html)). It generates 2D images from files, which can be a bit daunting to understand, but are key to grasp redundancy patterns. `fv` plots the distribution of string matches of length 1, 2, 4, and 8 by location in file (x axis) and distance backwards to the previous match (y axis, log scale).  The length of the match is color coded: black=1, red=2, green=4, blue=8.

For a first example, let's have a look at enwik8 : 
![enwik8](http://mattmahoney.net/dc/e8.png)

Now for comparison, let's have a look at silesia :
![silesia](http://mattmahoney.net/dc/silesia.png)

Distribution differences between files inside the Silesia corpus are striking. One can immediately "see" where internal files start and stop.
Among them, 2 seem to show the same pattern as enwik8 :

- dickens : Collected works of Charles Dickens,  10 192 446 bytes
- webster : The 1913 Webster Unabridged Dictionary,  41 458 703 bytes

Unsurprisingly, both are human-redacted text files.

Compare this with latest segment

- xml : Collected XML files,  5 345 280 bytes

Patterns are completely different.

`enwik` produces a lot of _small_ matches at small distances. This is perfect to **study** and show off strengthes of dynamic context modeling, which was the entire point. Even Matt brands it as ["_to promote AI research_"](http://mattmahoney.net/dc/).

But in real life, almost all compression workload are completely different : they tend to produce longer matches, at longer but more regular distances, and lesser frequence. Such characteristics are less interesting for context modeling, which can't flex their muscle with all this "noise" around.

In conclusion, `enwik8` can be considered a reference for " human-redacted text files, but that's all, and it's a diminishing part of content to compress.

>  you only benchmark the file silesia, and you omit text files like enwik8 or database files etc.

The good point about the silesia corpus is that it consists of several files, and the university made a serious effort at trying to mix multiple representative and freely available file types. They are also [listed on the Squash benchmark homepage](http://quixdb.github.io/squash-benchmark/) by the way. As we have just seen, _there are some text files_, quite a large amount by the way (almot 30%!), but also source codes, XML schematics, HTML files, standalone executable and library, image, and yes even some _database contents_. So it is much more diverse corpus. Consequently, it's more representative of some "all-around" compression workload.

There are some other corpus which are worthy of mentioning, most importantly the Calgary corpus, and the Canterbury corpus. They became reference a long time ago, and are a bit small and outdated by now (but still interesting).
Lesser known, there was also the [Compression Rating corpus](http://compressionratings.com/documents.html#corpus). It's very large and diverse, but obviously has shortcomings of its own. It was primarily designed to show off the strengthes of the author's compression algorithm, and is therefore a bit twisted in its favor (for example, it has some large audio file in .wav format, because his own algorithm features a very good delta filter for such file). Nonetheless, overall, it's a relatively good corpus, as the author made a good effort at trying to represent many different situations.

In the end, the Silesia Corpus is still to be preferred : it's more recent, it's large enough but not too much (the cheer size of the CompressionRating corpus make it difficult to benchmark it regularly on any target), and the legal distribution status of silesia files is very clean. Better be in the clear, since the goal is to allow anyone to reproduce results safely on its own system.

**Comment by Cyan4973** [comment]
Selecting a benchmark methodology
---------------------------------------------------

Having a reference corpus is only one part of the equation.
Now comes the difficult question of what to do with it, and what kind of result to produce.

> What I don't understand is that you use fsbench (...) I know you're aware of the fantastic benchmark by @nemequ which can be seen here, it's probably the best there is today (lots of platforms, files, codecs)... 

Initially, the situation was a bit murcky : there were many public benchmarks over internet, applying their own methodology, sometimes questionable, rarely replicable. Better tools and methodologies came from programmers which were themselves compression experts. This was a direct consequence of their understanding of the field. But unsurprisingly, it also resulted in subtle (and changing) selection criteria, which could be designed to favor their own work. It's not difficult to make a cheated "fair-looking contest" : take an ape and a fish, contest them on which one climb the tree better, or swim better. Get your winner.

With this past in mind, a choice was made at the early beginning of LZ4 : provide results based on verifiable open-sourced benchmark program _not controlled by an author of a compression library_. This property is critical, otherwise I could simply provide results based on my own benchmark programs, introducing bias and suspicions. (Note that there is always a bias, we are just trying to limit its scope as best as possible).

Consequently, I only link to public versions that anyone can download and run. I don't link to "my own little modification", nor other's one, since it would violate the rule above.

The latest public version of fsbench is 0.14.3. Its author, Maciej Adamczyk, has selected to make it a very generic benchmark platform, so there are a tons of things that can be benchmarked with it, many of them not relevant for data compression.

The most widespread fast compression codec are properly supported. It's not a secret, they are Snappy, LZO, QuickLZ and LZ4. LZF belongs to this list too, as it used to be popular when there was less alternatives. It also used to be part of the table, it's just an overlook if a recent format update dropped it.  I didn't even noticed, given it consistently ranked last in previous benchmarks. Nonetheless, it's back.

The other compression algorithms you mention are either not part of fsbench, or have not achieved any significant traction yet within the programming community.

Being part of fsbench can be a problem for recent algorithms, since it's no longer actively supported. But fortunately, there are some better perspectives for the future.

That's probably one of the most important point : @nemequ work on compression benchmark is top notch. I expect his benchmark system to become a reference. Consequently it's also very likely to become _the_ reference benchmark of LZ4 in the foreseeable future. At which stage, obviously, it will be linked for people willing to dig further into benchmarking numbers.

I also support the view of Evan : a README is not a fully detailed benchmark evaluation. You need a lot more space and a lot more work to provide such insight. The README is merely an attempt to position the algorithm somewhere between hopefully known reference points. That's why I see tools such as [@nemequ squash benchmark](https://quixdb.github.io/squash-benchmark/?dataset=cp.html&machine=s-desktop) as important and complementary.

That being said, it's still early days for Squash benchmark, and there are some early issues to solve. I'm sure they will be sorted out, it just needs a bit of worktime. As a simple example, I've been unable to reproduce a squash benchmark on my own test system. It's certainly not unsolvable (there are some dependency issues, linking issues, directory issues, codec issues, etc.), but it simply means it's not ready as a simple tool to download/compile by non-expert programmers. It's not a big deal, I know the current code is mostly a PoC (there is not even an official release). I don't want to bother Evan yet with this, since he is working on a next version. When it will be ready, I'll make some new tests, and contribute the best I can to his project.

Also, it's still too soon to link current results, as they can, at times, prove inconsistent. For example [these ones] (https://quixdb.github.io/squash-benchmark/?dataset=cp.html&machine=s-desktop), where a few fast compressors including LZ4 score an abysmal decompression speed < 20 MB/s, which can't be correct by a few order of magnitude. Once again, it's not reason to worry yet. As said, it's normal for a first version to have a few issues to sort out. It can be a fluke, a minor change, an intermediate bug, well whatever. So let's wait for the next version, and we'll start building on this future one.

**Comment by Cyan4973** [comment]
Selecting reference comparison algorithms
-----------------------------------------------------------

OK, so I believe it is the part which most interests you.


> It's nice to add zlib as a reference but is it really useful ? 

Yes it is. zlib is a reference. Basically, it is _the_ absolute reference, I can't imagine a data compression benchmark without it, even if it targets a different trade off (more compression, less speed).

In the case of LZ4, it's also quite useful as it shows `LZ4 HC` can reach `zlib - 1` compression level while preserving decompression speed, which is the core of the message.


> it is very simple to add a line with another codec, especially when the codec in question is a direct alternative to your library.
> > The problem is where to stop.
> 
> Yes and no... there aren't many open source libraries that "compete" in the same areas.

I guess you underestimate the number of alternatives available in your category. You could have listed Yappy, or crush, lzmat, aplib, lzbw, fastlz, lzv, szip, lzp2, lzrw, etc. Well, I guess no one can pretend to know a complete list, as it's constantly evolving.

Furthermore, many of these algorithms have not reached "production" status, and never will, sacrificing portability, safety or correctness in favor of quick roll out or performance, so they are mostly short interesting experiments. 

Therefore, there is a need to draw a line somewhere. In my case, this is all the more necessary as I'm regularly receiving requests to include some new experimental algorithm on the LZ4 front page.

I ended with 2 simple rules :

- Algorithm should be part of the benchmark suite (which content I don't control)
- Algorithm should be "production ready", with references

Regarding density, I have been tempted to include it, as it occupies an unusual trade-off on the compression/speed pareto frontier. In the end, I realised it did not satisfied any of above conditions, and so ruled against it.
Point 1 will likely be solved by moving to @nemequ squash benchmark. As to point 2, it's now totally up to you.

**Comment by gpnuma** [comment]
Yann, thanks for your answers as it's always constructive to hear someone's opinion.

**Selecting a Reference data corpus**
Let me start by praising your answer which is very complete and contains some easy to understand graphics. I completely agree with you for the most part.

> In the end, the Silesia Corpus is still to be preferred : it's more recent, it's large enough but not too much (the cheer size of the CompressionRating corpus make it difficult to benchmark it regularly on any target), and the legal distribution status of silesia files is very clean. Better be in the clear, since the goal is to allow anyone to reproduce results safely on its own system.

Honestly I quite like silesia and I actually use it to benchmark density, that's not the point I am trying to make. What I mean is that it's a compact file consisting obviously of smaller files, and depending on their data size distribution it can "hide" flaws or lack of performance in some areas, like *lz4 fast* on text files.
That's why I would advise in showing performance with another well known dataset, and as you said Matt Mahoney's enwik8 has gained a lot of popularity in recent years, even if it's text specific. That's what I'm using anyways.
Paradoxically enwik8, by using a single type of data, is also a good indicator of how well one's algorithms performs in relation to text compression, but it's also its flaw : it doesn't show any indication whatsoever about other data types.
But at least, with enwik8 you can say that your codec performs well or doesn't in regards to text compression. With silesia, you can't say anything of the sort, the only conclusion you get is that your codec is good with the silesia file itself, period. That's its main limitation for me.

**Selecting a benchmark methodology**

> The other compression algorithms you mention are either not part of fsbench, or have not achieved any significant traction yet within the programming community.

Well if you're talking about density specifically, you're misinformed about fsbench ... its repository is located [here](https://chiselapp.com/user/Justin_be_my_guide/repository/fsbench/tree?ci=tip) and it includes a version of density (take a look at the codecs directory).

As for traction ? I am aware of projects using density, but they are in their early stages and have not published releases so far. That's completely understandable because the library itself is still in beta, does that mean no one is interested ? I don't think so.

> With this past in mind, a choice was made at the early beginning of LZ4 : provide results based on verifiable open-sourced benchmark program not controlled by an author of a compression library. This property is critical, otherwise I could simply provide results based on my own benchmark programs, introducing bias and suspicions. (Note that there is always a bias, we are just trying to limit its scope as best as possible).

Now that's more disturbing, because it seems to me that you're doing *exactly the opposite*.
The latest version in m^2's repository features LZ4 r127, and the numbers you publish (once again without giving the sources) are based on r129, so you have obviously tweaked m^2's sources to include r129. As I see it you are, as a compression library author, actually modifying his benchmark.
If you want an up-to-date version with LZ4 r129 included, sources are [here](https://github.com/gpnuma/fsbench-density.git) - they are available to all but I have updated the libraries.

> Also, it's still too soon to link current results, as they can, at times, prove inconsistent. For example these ones, where a few fast compressors including LZ4 score an abysmal decompression speed < 20 MB/s, which can't be correct by a few order of magnitude. Once again, it's not reason to worry yet. As said, it's normal for a first version to have a few issues to sort out. It can be a fluke, a minor change, an intermediate bug, well whatever. So let's wait for the next version, and we'll start building on this future one.

Couldn't disagree more here. Even if 1% of the presented results are doubtful, the vast majority of it is.
It is consistent across platforms of the same type (x86 / ARM) and the type of files used (text/db etc.).
Even though I'd like Squash to test density compiled with Clang/LLVM (its target compiler), I couldn't support @nemequ 's results more.
They completely concur with what I've tested on my platform with GCC.
To me, not linking to the best and most up-to-date benchmark there is, just because some corner cases look suspicious, is either bad will or a mistake.

**Selecting reference comparison algorithms**

> Yes it is. zlib is a reference. Basically, it is the absolute reference, I can't imagine a data compression benchmark without it, even if it targets a different trade off (more compression, less speed).

OK, why not.

> I ended with 2 simple rules :
Algorithm should be part of the benchmark suite (which content I don't control)
Algorithm should be "production ready", with references

In the specific case of density, it's part of both fsbench and squash. It's not production ready yet, although this will likely happen later this year.
But then, we both agree that compression libraries are targeted for developers and not end users right ? Then why make the assumption that developers only look for production-ready libraries ? That is very far from being the reality !

**Comment by nemequ** [comment]
> It's not a big deal, I know the current code is mostly a PoC (there is not even an official release).

Actually there has been one release (0.5).  0.6 is [*almost* ready](https://github.com/quixdb/squash/milestones/0.6) (the only non-trivial thing on that list is quixdb/squash#97, and that's mostly done—just going through updating the plugins to use the new API now).  I anticipate finishing everything up this week, re-running the benchmark next week, and releasing 0.6 as soon as it finishes.

> That being said, it's still early days for Squash benchmark, and there are some early issues to solve. I'm sure they will be sorted out, it just needs a bit of worktime. As a simple example, I've been unable to reproduce a squash benchmark on my own test system. It's certainly not unsolvable (there are some dependency issues, linking issues, directory issues, codec issues, etc.), but it simply means it's not ready as a simple tool to download/compile by non-expert programmers. It's not a big deal, I know the current code is mostly a PoC (there is not even an official release).

I recently made some changes to make building a lot simpler.  Most significantly there are now copies of all plugins in Squash, so dependencies shouldn't be a problem (you need a C compiler, a C++ compiler, CMake, and Ragel).  By default Squash will still use system libraries where available, but it will fall back on the local copy otherwise; you can pass `--disable-external` to always use the local copies.

I also wrote [a wrapper for CMake which presents an autotools-style interface](https://github.com/nemequ/configure-cmake/) which is included in Squash, so you can just do an autotools-style `./configure && make && make install` (or, also like autotools, if you're using git instead of a release `./autogen.sh && make && make install`).  You can pass it things like `--prefix`, `--libdir`, etc., just like autotools… `./configure --help` will provide a complete list.

There are directions for [building squash](https://github.com/quixdb/squash/blob/master/docs/building.md) included in the documentation; it should be pretty easy now.  If you still have any problems please file an issue… I really want building Squash to be easy, so bug reports would be *very* welcome.

> Also, it's still too soon to link current results, as they can, at times, prove inconsistent. For example these ones, where a few fast compressors including LZ4 score an abysmal decompression speed < 20 MB/s, which can't be correct by a few order of magnitude. Once again, it's not reason to worry yet. As said, it's normal for a first version to have a few issues to sort out. It can be a fluke, a minor change, an intermediate bug, well whatever. So let's wait for the next version, and we'll start building on this future one.

Yes, that's troubling…  I've filed an issue (quixdb/squash-benchmark#17).  I've seen anomalous results occasionally before, but so far I haven't had much luck triaging them…

> I don't want to bother Evan yet with this, since he is working on a next version. When it will be ready, I'll make some new tests, and contribute the best I can to his project.

Please don't be shy about filing issues.  I don't have any problem with the issue tracker for squash, the benchmark, or the corpus filling up—Constructive criticism is fantastic, and anything which would help me make Squash better would be greatly appreciated.  If you think it's too early to be concerned about something it can still go in the issue tracker; it may just be a while until I get around to fixing it.  I'll also try to be good about assigning milestones so you have a basic idea of what kind of time frame I think would be appropriate.

**Comment by Cyan4973** [comment]
> Please don't be shy about filing issues. I don't have any problem with the issue tracker for squash, the benchmark, or the corpus

Sure, you are right.

I was just not very enthusiastic on sending a weak issue such as "I don't agree with this result".
It seems proper to study it a bit before complaining, and to do this, I needed the ability to reproduce tests.
With 0.6 almost there, I guess we are at the right point to start this process.

**Comment by Cyan4973** [comment]
I guess we learned all we could from this discussion, so let's move on

**Comment by gpnuma** [comment]
I don't want to sound unpleasant but as soon as some unfavorable questions are asked (read my previous posts) they actually get no answers.

After all everything looks great, the README shows that LZ4 is faster than an incomplete selection of slower codecs, now that's very interesting ! And ... oh yeah I forgot... it's also faster than zlib.

Seriously ?

**Comment by lpsantil** [comment]
Maybe README should just refer to a 3rd Party Benchmarker [0], even if some
the program versions are not up to date.

                Compression                      Compressed size
Decompresser  *Total size*   Time (ns/byte)Program
<http://mattmahoney.net/dc/text.html#about>           Options
             enwik8      enwik9     size (zip)   enwik9+prog  Comp
Decomp  Mem Alg Note <http://mattmahoney.net/dc/text.html#notes>
-------           -------                     ----------  -----------
-----------  -----------  ----- -----  --- --- ----

lz4 <http://mattmahoney.net/dc/text.html#3800> v1.2          -c2
                  42,870,164  379,999,522     49,128 x  380,048,650
 91     6   20 LZ77 26

gzip <http://mattmahoney.net/dc/text.html#3226> 1.3.5        -9
                  36,445,248  322,591,995     38,801 x  322,630,796
101    17  1.6 LZ77

lzop <http://mattmahoney.net/dc/text.html#3663> v1.01        -9
                  41,217,688  366,349,786     54,438 x  366,404,224
289    12  1.8 LZ77

snappy <http://mattmahoney.net/dc/text.html#5277> 1.0.1
                  58,350,605  527,772,054     23,844 s  527,795,898
 25    12  0.1 LZ77 26

lzf <http://mattmahoney.net/dc/text.html#4068> 1.02          cx
                  45,198,298  406,805,983     48,359 x  406,854,342
 68   *2.2*  151 LZ77 68

QuickLZ <http://mattmahoney.net/dc/text.html#4106> 1.30b     (quick3)
                  46,378,438  410,633,262     44,202 x  410,677,464
 48    12    3 LZ77

FastLZ <http://mattmahoney.net/dc/text.html#4930> Jun 12 2007
                  54,658,924  493,066,558      7,065 xd 493,073,623
 18    13    1 LZ77

bzip2 <http://mattmahoney.net/dc/text.html#2540> 1.0.2       -9
                  29,008,736  253,977,839     30,036 x  254,007,875
379   129    8 BWT

xz <http://mattmahoney.net/dc/text.html#1973>
5.2.1--lzma2=preset=9e,dict=1GiB,lc=4,pb=0 24,703,772  197,331,816
36,752 xd 197,368,568   5876    20 6000 LZ77 73

lzham <http://mattmahoney.net/dc/text.html#2024> 1.0         -d29 -x
                  25,002,070  202,237,199    191,600 s  202,428,799
1096   *6.6* 7800 LZ77 70



[0] http://mattmahoney.net/dc/text.html

**Comment by Cyan4973** [comment]
> Maybe README should just refer to a 3rd Party Benchmarker [0], even if some
the program versions are not up to date.

That's actually the plan,
although the candidate target is @nemequ Squash benchmark
when it will be ready (i.e. likely soon).

For reasons explained above, I'm not inclined using Matt's `enwik` as a reference corpus, it's too far away from real-life compression workload.

**Comment by nemequ** [comment]
I'm finally finishing up the update to the next version (I'm on the last two machines, I expect them to finish sometime tomorrow).  Current plan is to release a new version of Squash and the benchmark late Sunday (-0700) or early Monday, but I thought you might want [a sneak peek](https://github.com/quixdb/squash-benchmark-web).  It includes some new charts, and of course lz4 has been updated to the latest version, along with all the other libraries.

I'm trying to be more careful about making sure nothing else is running while I run the benchmarks, so hopefully the results are a bit more stable, but if anyone notices anything odd please let me know!

Sorry it took so long to get this done—work got *really* crazy for a while :/

**Comment by Cyan4973** [comment]
Thanks very much for the sneak peak.
I'm really eager to test your new benchmark suite on a few machines around.

I've looked into preliminary data, they look pretty good to me.
I also like the new graphics / display capabilites, 
they offer some truly great perspectives.

Regards

**Comment by nemequ** [comment]
Just released new versions of both the library and the benchmark.

> I'm really eager to test your new benchmark suite on a few machines around.

Just a warning, it takes like 21 hours to run on my Xeon E3-1225 v3 desktop.

> I've looked into preliminary data, they look pretty good to me.
> I also like the new graphics / display capabilites,
> they offer some truly great perspectives.

I thought you would like the transfer + processing chart; LZ4 performs really well there.

**Comment by Cyan4973** [comment]
> Just a warning, it takes like 21 hours to run on my Xeon E3-1225 v3 desktop.

Is there a way to select algorithms and samples to benchmark. or is this a hard-coded list ?

**Comment by gpnuma** [comment]
Hey Evan,

Thanks a lot for the sneak peak, that's a great piece of work ! The new graphics are very interesting indeed !
I've been away for quite a while now, but can't wait to test the upcoming 0.13 version of density on your benchmark :)

**Comment by nemequ** [comment]
> Is there a way to select algorithms and samples to benchmark. or is this a hard-coded list ?

You can pass the executable a list of files to run the benchmark on, and you can pass a specific codec with the -c flag.  Here is the help:

```
Usage: ./benchmark [OPTION]... FILE...
Benchmark Squash plugins.

Options:
	-h            Print this help screen and exit.
	-c codec      Benchmark the specified codec and exit.
	-o outfile    JSON output file.
```

However, the web site is currently hard coded for the defaults.  It wouldn't be hard to change—you would just need to tweak the [datasets](https://github.com/quixdb/squash-benchmark-web/blob/master/squash-benchmark.js#L71) and [machines](https://github.com/quixdb/squash-benchmark-web/blob/master/squash-benchmark.js#L219) list.  See also quixdb/squash-benchmark-web#4
